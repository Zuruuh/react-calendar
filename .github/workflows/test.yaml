name: Tests and static analysis

on:
  push:
    branches: ["dev"]
  pull_request:
    branches: ["dev"]

jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - name: ðŸ›’ Checkout branch
        uses: actions/checkout@v4

      - name: âœ¨ Install bun
        uses: oven-sh/setup-bun@v1
        with:
          bun-version: 1.0.25

      - uses: actions/cache@v4
        name: ðŸ“¦ Setup deps cache
        with:
          path: node_modules
          key: ${{ runner.os }}-modules-${{ hashFiles('**/bun.lockb') }}
          restore-keys: |
            ${{ runner.os }}-modules-

      - name: âž• Install dependencies
        run: bun install --frozen-lockfile

      - name: ðŸš¨ Check formatting
        run: bun run format

      - name: ðŸš¨ Lint
        if: always()
        run: bun run lint

      - name: ðŸ§ª Tests
        if: always()
        run: bun test
