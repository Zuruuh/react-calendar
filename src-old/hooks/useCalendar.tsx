import { type FC, ReactNode, useCallback } from 'react';
import type {
  CalendarPlugin,
  CalendarPluginDefinition,
  CalendarPlugins,
} from '../plugin';
import day, { type Dayjs } from 'dayjs';
import type {
  CalendarContext,
  CalendarState,
} from '../context/CalendarContext';

export type Calendar<TPlugins extends CalendarPlugins> = {
  calendar: {
    Root: FC<CalendarProps>;
  };
} & MergedPluginProps<TPlugins, 'rootState'>;

type MergedPluginProps<
  T extends CalendarPlugins,
  Key extends keyof CalendarPluginDefinition,
> = T extends [infer First, ...infer Rest]
  ? First extends CalendarPlugin<infer FirstProps>
    ? Rest extends CalendarPlugins
      ? FirstProps[Key] & MergedPluginProps<Rest, Key>
      : FirstProps[Key]
    : {}
  : {};

export type CalendarOptions<TPlugins extends CalendarPlugins> = {
  viewedDate?: Dayjs;
  dayjs?: () => Dayjs;
} & MergedPluginProps<TPlugins, 'rootConfiguration'>;

export type CalendarProps = {
  children: ReactNode;
};

export function useCalendar<
  const TPlugins extends Readonly<CalendarPlugins> = readonly [],
>({
  plugins,
  dayjs = () => day(),
  viewedDate,
  ...props
}: { plugins: TPlugins } & CalendarOptions<
  TPlugins extends readonly [...infer U] ? U : never
>): Calendar<TPlugins extends readonly [...infer U] ? U : never> {
  // function RootComponent({ children }: CalendarProps) {
  //   const dayFactory = useCallback(
  //     () => dayjs().utc(true).second(0).minute(0).hour(12),
  //     [dayjs],
  //   );
  //
  //   const baseState: CalendarState & Record<string, unknown> = {
  //     dayjs,
  //     viewedDate: viewedDate ?? dayFactory(),
  //     ...props,
  //   };
  //
  //   const state = {
  //     ...baseState,
  //     ...plugins
  //       .map((plugin) =>
  //         plugin.calendarHook !== undefined
  //           ?
  //             plugin.calendarHook(baseState) : {},
  //       )
  //       .reduce((acc, curr) => ({ ...acc, ...curr }), {}),
  //   };
  //
  //   return (
  //     <>{children}</>
  //   );
  // }

  return {
    calendar: {
      Root: Calendar,
    },
  };
}

const Calendar: FC<CalendarProps> = ({ children }) => {
  return (
    <CalendarContext.Provider value={{}}>{children}</CalendarContext.Provider>
  );
};
