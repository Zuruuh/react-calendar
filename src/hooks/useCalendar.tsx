import {
  type ReactNode,
  useMemo,
  Dispatch,
  SetStateAction,
  useCallback,
} from 'react';
import type { Dayjs, ManipulateType } from 'dayjs';

export type Calendar = {
  weeks: Array<WeekState>;
  controller: CalendarController;
};

export interface WeekState {
  startWeek: number;
  endWeek: number;
  days: Array<DayState>;
}

interface DayState {
  date: Dayjs;
  outsideViewedMonth: boolean;
}

export type CalendarOverlap = 'overlap' | 'no-overlap';

export type CalendarOptions = {
  viewedDate: Dayjs;
  setViewedDate: Dispatch<SetStateAction<Dayjs>>;
  dayjs: () => Dayjs;
  overlap: CalendarOverlap;
};

export type CalendarController = (count: number, unit: ManipulateType) => void;

export type CalendarProps = {
  children: ReactNode;
};

export function useCalendar(props: CalendarOptions): Calendar {
  const controller = useCallback<Calendar['controller']>(
    (count, unit) => {
      props.setViewedDate(props.viewedDate.add(count, unit));
    },
    [props.viewedDate, props.setViewedDate],
  );

  const startOfMonth = props.viewedDate
    .startOf('month')
    .hour(0)
    .minute(0)
    .second(0);
  const endOfMonth = props.viewedDate
    .endOf('month')
    .hour(23)
    .minute(59)
    .second(59);

  const rawWeeks = useMemo(
    () => Array.from(generateWeeksBasedOnOverlap(startOfMonth, props.overlap)),
    [startOfMonth.toString(), props.overlap],
  );
  const weeks: Array<WeekState> = useMemo(
    () =>
      Array.from({ length: rawWeeks.length }).map((_, i) => {
        const week = rawWeeks[i]!;
        const days: Array<DayState> = [];

        for (let j = 0; j < 7; j++) {
          let date: Dayjs;
          switch (props.overlap) {
            case 'overlap':
              date = startOfMonth
                .startOf('week')
                .add(i, 'weeks')
                .add(j, 'days');
              break;
            case 'no-overlap':
              date = startOfMonth.add(i, 'weeks').add(j, 'days');
              break;
          }

          days.push({
            date,
            outsideViewedMonth:
              date.isBefore(startOfMonth) || date.isAfter(endOfMonth),
          });
        }

        return { startWeek: week[0], endWeek: week[1] ?? week[0], days };
      }),
    [rawWeeks],
  );

  return useMemo(
    () => ({
      controller,
      weeks,
    }),
    [controller, weeks],
  );
}

/**
 * @internal
 */
function overlapLoopCheck(
  i: number,
  startOfMonth: Dayjs,
  referenceMonth: number,
): boolean {
  if (i === 0) {
    return true;
  }

  const date = startOfMonth.add(i, 'weeks');

  return [date.startOf('week').month(), date.endOf('week').month()].includes(
    referenceMonth,
  );
}

/**
 * @internal
 */
export function* generateWeeksBasedOnOverlap(
  referenceDate: Dayjs,
  overlap: CalendarOverlap,
): Generator<[number, null | number]> {
  const startOfMonth = referenceDate.startOf('month');

  switch (overlap) {
    case 'overlap':
      for (
        let i = 0;
        overlapLoopCheck(i, startOfMonth, referenceDate.month());
        i++
      ) {
        yield [startOfMonth.add(i, 'week').startOf('week').week(), null];
      }

      break;
    case 'no-overlap':
      for (let i = 0; i < Math.ceil(startOfMonth.daysInMonth() / 7); i++) {
        yield [
          startOfMonth.add(i, 'week').week(),
          startOfMonth.add(i + 1, 'week').week(),
        ];
      }

      break;
  }
}
