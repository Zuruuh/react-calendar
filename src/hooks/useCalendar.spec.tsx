import { expect, test } from 'bun:test';
import { type WeekState, useCalendar } from './useCalendar';
import { render } from '@testing-library/react';
import dayjs from 'dayjs';

const mockedDayjs = () => dayjs().year(2024).month(1).date(27);
const today = mockedDayjs();

function formatWeeksForSnapshot(weeks: Array<WeekState>): unknown {
  return weeks.map((week) => ({
    ...week,
    days: week.days.map((day) => ({
      ...day,
      date: day.date.format('D-MM-YYYY'),
    })),
  }));
}

test('Calendar with default overlap mode', () => {
  const MyComponent = () => {
    const { weeks } = useCalendar({
      initialViewedDate: today,
      dayjs: mockedDayjs,
      overlap: 'overlap',
    });

    expect(formatWeeksForSnapshot(weeks)).toMatchSnapshot();

    return <input />;
  };

  render(<MyComponent />);
});

test('Calendar with no-overlap mode', () => {
  const MyComponent = () => {
    const { weeks } = useCalendar({
      initialViewedDate: today,
      dayjs: mockedDayjs,
      overlap: 'no-overlap',
    });

    expect(formatWeeksForSnapshot(weeks)).toMatchSnapshot();

    return <input />;
  };

  render(<MyComponent />);
});
