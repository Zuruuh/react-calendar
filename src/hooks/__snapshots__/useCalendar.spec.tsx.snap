// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Calendar with default overlap mode 1`] = `
[
  {
    "days": [
      {
        "date": "29-01-2024",
        "outsideViewedMonth": true,
      },
      {
        "date": "30-01-2024",
        "outsideViewedMonth": true,
      },
      {
        "date": "31-01-2024",
        "outsideViewedMonth": true,
      },
      {
        "date": "1-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "2-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "3-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "4-02-2024",
        "outsideViewedMonth": false,
      },
    ],
    "endWeek": 5,
    "startWeek": 5,
  },
  {
    "days": [
      {
        "date": "5-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "6-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "7-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "8-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "9-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "10-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "11-02-2024",
        "outsideViewedMonth": false,
      },
    ],
    "endWeek": 6,
    "startWeek": 6,
  },
  {
    "days": [
      {
        "date": "12-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "13-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "14-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "15-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "16-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "17-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "18-02-2024",
        "outsideViewedMonth": false,
      },
    ],
    "endWeek": 7,
    "startWeek": 7,
  },
  {
    "days": [
      {
        "date": "19-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "20-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "21-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "22-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "23-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "24-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "25-02-2024",
        "outsideViewedMonth": false,
      },
    ],
    "endWeek": 8,
    "startWeek": 8,
  },
  {
    "days": [
      {
        "date": "26-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "27-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "28-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "29-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "1-03-2024",
        "outsideViewedMonth": true,
      },
      {
        "date": "2-03-2024",
        "outsideViewedMonth": true,
      },
      {
        "date": "3-03-2024",
        "outsideViewedMonth": true,
      },
    ],
    "endWeek": 9,
    "startWeek": 9,
  },
]
`;

exports[`Calendar with no-overlap mode 1`] = `
[
  {
    "days": [
      {
        "date": "1-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "2-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "3-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "4-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "5-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "6-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "7-02-2024",
        "outsideViewedMonth": false,
      },
    ],
    "endWeek": 6,
    "startWeek": 5,
  },
  {
    "days": [
      {
        "date": "8-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "9-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "10-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "11-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "12-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "13-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "14-02-2024",
        "outsideViewedMonth": false,
      },
    ],
    "endWeek": 7,
    "startWeek": 6,
  },
  {
    "days": [
      {
        "date": "15-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "16-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "17-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "18-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "19-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "20-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "21-02-2024",
        "outsideViewedMonth": false,
      },
    ],
    "endWeek": 8,
    "startWeek": 7,
  },
  {
    "days": [
      {
        "date": "22-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "23-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "24-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "25-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "26-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "27-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "28-02-2024",
        "outsideViewedMonth": false,
      },
    ],
    "endWeek": 9,
    "startWeek": 8,
  },
  {
    "days": [
      {
        "date": "29-02-2024",
        "outsideViewedMonth": false,
      },
      {
        "date": "1-03-2024",
        "outsideViewedMonth": true,
      },
      {
        "date": "2-03-2024",
        "outsideViewedMonth": true,
      },
      {
        "date": "3-03-2024",
        "outsideViewedMonth": true,
      },
      {
        "date": "4-03-2024",
        "outsideViewedMonth": true,
      },
      {
        "date": "5-03-2024",
        "outsideViewedMonth": true,
      },
      {
        "date": "6-03-2024",
        "outsideViewedMonth": true,
      },
    ],
    "endWeek": 10,
    "startWeek": 9,
  },
]
`;
